### ISSUE TITLE
Codemirror text overflows editor

### ISSUE THREAD

**Issue by jasonLaster** [replayio/gecko-dev#292]
This is probably similar to #287, but easier because it doesn't have the unknown characters.

**STR:**

1. go to  https://replay.io/view?id=d9215cc2-6591-4d96-88f7-fe4024a5c0e1
2. play video

<img width="1126" alt="Screen Shot 2021-02-04 at 8 40 29 AM" src="https://user-images.githubusercontent.com/254562/106925156-aacb3780-66c4-11eb-91d4-1d5a6471dd15.png">

**Comment by bhackett1024** [comment]
This seems to start happening when using the gutter in CodeMirror, but not before:

https://replay.io/view?id=2316f416-46b6-47b3-9b78-951f8bb53085

**Comment by bhackett1024** [comment]
The precise STR here are:

1. Visit https://replay.io/view?id=02d1b4aa-16f4-4cba-a00d-2863f5c8b70a&point=0&time=0&hasFrames=false
2. Select devtools
3. Click on the first console message
4. Set a breakpoint on the paused line after the source opens.  There can't already be a breakpoint there.

I wrote a patch to allow recording processes to write JPGs of everything they are compositing to a folder.  This showed that the problem occurs while recording and not just replaying, and should make it easier to track down the problem.  From one of these runs, this is the last good paint:

![paint-355](https://user-images.githubusercontent.com/1062578/119202746-b1e93c00-ba4e-11eb-8ff7-f3cf67199bd9.jpg)

And the first bad paint is below.  It basically looks like the CodeMirror contents are overflowing everything, as the bug title says.  This seems like it could happen if the layers are being drawn in the wrong order, or if they are missing clipping information.

![paint-356](https://user-images.githubusercontent.com/1062578/119202954-2f14b100-ba4f-11eb-8ff3-09a02db87554.jpg)

I was wondering if this problem was happening because we are using a software compositor in the recording/replaying process, and a hardware compositor in the parent process.  I disabled acceleration and confirmed the parent process is using a software compositor, but the problem still continues.

**Comment by jasonLaster** [comment]
Thanks for the writeup @bhackett1024.

**Comment by bhackett1024** [comment]
I've figured this out.  tl;dr we're supposed to disable async pan/zoom (APZ) in recording/replaying tabs, but we aren't, and that's causing this rendering problem.

Tracking this down was pretty tricky.  As described above I wrote a patch to write JPGs of all renders to a directory, and then extended this to write JPGs at intermediate points during each render, i.e. before/after each layer is composited to the record/replay graphics draw target.  With this I was able to determine the first layer composite which overflowed, and see the clipping information used to render that layer.  By finding the layer handle (an integer ID used in IPDL messages) for this layer I matched this up with the corresponding layer in the parent and the clipping information used to render it.  These were different, and I tracked that back to a call to SetShadowClipRect() which we were making in the parent process but not the recording process.  That call happened because of some APZ state being set which wasn't supposed to be set.  Disabling APZ globally fixed the rendering.

The full patch is in the gist below.  I'll create PRs for the JPG rendering (which is generally useful) and the fix itself, and we can refer to this patch if we run into other problems in the future.

https://gist.github.com/bhackett1024/084ce073f256d586110aa6bc19c6b49c
